src/jit/lowerxarch.cpp

```diff
@@ -1174,13 +1174,13 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
     else
     {
         MakeSrcContained(tree, shiftBy);
-    }
 
-    // Codegen of shift oper sets ZF and SF flags.
-    // Note that Rotate Left/Right instructions don't set ZF and SF flags.
-    if (tree->OperIsShift())
-    {
-        tree->gtFlags |= GTF_ZSF_SET;
+        // Codegen of shift oper sets ZF and SF flags provided shift count is not zero.
+        // Note that Rotate Left/Right instructions don't set ZF and SF flags.
+        if (tree->OperIsShift() && (shiftBy->gtIntConCommon.IconValue() != 0))
```

Comment:
Yes, you are right, depending on the size of the operand the shift-count gets narrowed by the hardware by making upper bits.  For e.g. while shifting 32-bit operands, upper 3 bits of shift count is masked by hardware to derive shift count.  If the resulting shift count is zero, then it won't modify the flags.

Given the urgency of this fix, I am going to disable this optimization for shift operations with a CQ issue to implement correctly.